### ISSUE TITLE
[NEW] Allow building with LuaJIT

### ISSUE THREAD

**Issue by secwall** [madolson/placeholderkv#1229]
**The problem/use-case that the feature addresses**

In some cases Valkey users put some not-trivial logic in lua scripts. LuaJIT is a bit more performant than Lua (but has some limitations). We could add ability to build Valkey with LuaJIT instead of Lua.

**Description of the feature**

Add a configuration option `USE_LUAJIT`, that would build Valkey with LuaJIT instead of Lua. Proof-of-concept is here: https://github.com/secwall/valkey/tree/luajit. Most tests are passing (large strings are not possible in LuaJIT, so 3 GB cjson test fails). For now only x86_64 arch is sandboxed, but arm64 seem to work (at least on my Macbook M1).

**Alternatives you've considered**

I've tried to take a look at https://luau.org. But it's not a drop-in replacement of Lua and requires some hacks to implement it properly.

**Additional information**

I've tried to run a benchmark from #1168.
Unstable:
```
  throughput summary: 9683.35 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        5.156     0.112     5.207     6.423     7.967    10.815
```
LuaJIT:
```
  throughput summary: 19650.22 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        2.536     0.088     2.535     3.055     3.647     6.295
```
CPU: `Intel(R) Xeon(R) CPU E5-2660 v4 @ 2.00GHz`, no save, io-threads=9.

**Comment by efimkin** [comment]
wow!

**Comment by zuiderkwast** [comment]
Great! Are the benchmarks the same as in #1168 where you work with values of size 100KB? Can you test it with a more common workload? I believe LuaJIT is faster, but it would be interesting to see some more benchmark comparisons:

* Small or large values
* More values, such as multiple keys or multiple fields from hash or set type keys
* A few lines or many lines of Lua code

LuaJIT is compatible with Lua 5.1, which is what we need. It has been discussed in Redis a few years ago but it was not accepted.

If we add some option to build Valkey without Lua (#1204), we can also make it possible to choose a different Lua implementation.

We have some patches to our vendored Lua verions, to make it more sandboxed and safe(?). Do we need to vendor LuaJIT and add similar patches for that? It would be nice if we can avoid patching it and vendoring it.

**Comment by secwall** [comment]
> Are the benchmarks the same as in https://github.com/valkey-io/valkey/pull/1168 where you work with values of size 100KB?

Yes, benchmark is the same (large strings, same lua scripts, same way of running `valkey-benchmark`)

> I believe LuaJIT is faster, but it would be interesting to see some more benchmark comparisons:
> Small or large values
> More values, such as multiple keys or multiple fields from hash or set type keys
> A few lines or many lines of Lua code

Ok, I'll try to benchmark with these examples. Any other ideas of benchmarks are also welcome.

> We have some patches to our vendored Lua verions, to make it more sandboxed and safe(?). Do we need to vendor LuaJIT and add similar patches for that? It would be nice if we can avoid patching it and vendoring it.

I'm not a security expert, but I've tried to sandbox luajit as much as vendored lua is sandboxed. Quick link to patches: https://github.com/secwall/valkey/tree/luajit/deps/luajit/patches. The most interesting from security perspective are 0003, 0005, and 0007. I've mentioned it in first message, but let me clarify it again: only x86_64 is sandboxed for now (because it requires a bit of assembly knowledge to implement readonly tables).

**Comment by zuiderkwast** [comment]
Yes, I saw the patches now. I guess we need the sandboxing patches.

I wonder if it would be possible to upstream these patches. Improved sandboxing would be good for LuaJIT in general, I hope... But it would be a long-term project to do that.

We have a few options:

1. Replace vendored patched Lua with vendored patched LuaJIT. Is LuaJIT fully compatible in all aspects?
2. Add vendored patched LuaJIT to our repo, but keep regular Lua too and compile with regular Lua by default.
3. Avoid adding vendored code to the repo. Just add the patches and provide some way (instructions, scripts) to patch and build LuaJIT.

It's not perfect to add a lot of vendored code, but if we delete the other Lua, then maybe it's a net sum zero of vendored code. So I don't like option 2, but maybe 1 or 3.

@valkey-io/core-team @hpatro WDYT?

**Comment by secwall** [comment]
> Replace vendored patched Lua with vendored patched LuaJIT. Is LuaJIT fully compatible in all aspects?

Unfortunately, no. LuaJIT is not fully Lua-compatible. One example is: large strings. LuaJIT max string length is 2GB.
So it seems that we need option 3. I've added instruction to update vendored luajit to `deps/README.md`, so it's possible to automate getting luajit and patching it during the build process. Could we assume that `git` and internet access are available during the build process?

**Comment by zuiderkwast** [comment]
> Unfortunately, no. LuaJIT is not fully Lua-compatible. One example is: large strings. LuaJIT max string length is 2GB. So it seems that we need option 3.

This is a point for discussion and core team decision. I can't make this decision alone. Max string length 2GB would be a breaking change, but some breaking changes are acceptable in some cases, if there is a good reason. Maybe this can be acceptable, or maybe not. Anyway, option 3 seems like a safer choice to start with.

> I've added instruction to update vendored luajit to `deps/README.md`, so it's possible to automate getting luajit and patching it during the build process. Could we assume that `git` and internet access are available during the build process?

I know some companies require an offline build process and verified versions of deps. If we have the patches, it can be easy to adjust the process using some variables or parameters to the script, I guess. For example, if the user provides a path to the LuaJIT source code, we can use that and otherwise we can download it. WDYT?

For patching, we can use `git`. (Does `git apply` work if we're not in a git repo? If not, then `patch`?).

**Comment by secwall** [comment]
In my opinion Valkey should stay as compatible with open source version of Redis (7.2) as possible to simplify the migration process. At least for now.

> For example, if the user provides a path to the LuaJIT source code, we can use that and otherwise we can download it.

Good idea, I'll try to adjust makefile in `deps/` dir to use provided path. Does `LUAJIT_SOURCE_PATH` env variable name look sound?

> For patching, we can use git. (Does git apply work if we're not in a git repo? If not, then patch?).

Yes, it works:
```
secwall@secwall-osx[test] $ git st
fatal: not a git repository (or any of the parent directories): .git
secwall@secwall-osx[test] $ cat a.txt
ololo
azaza
secwall@secwall-osx[test] $ cat 0001-Commit.patch
diff --git a/a.txt b/a.txt
index 1b814a4..6732a43 100644
--- a/a.txt
+++ b/a.txt
@@ -1,2 +1,2 @@
 ololo
-azaza
+ololo
--
2.47.0
secwall@secwall-osx[test] $ git apply 0001-Commit.patch
secwall@secwall-osx[test] $ cat a.txt
ololo
ololo
```

**Comment by secwall** [comment]
I've updated my branch (https://github.com/secwall/valkey/tree/luajit) to use non-vendored variant. Now we checkout luajit if not found or copy if `LUAJIT_SOURCE_PATH` env variable is set.

I've also added support for sandboxing on aarch64, tests on Ubuntu Server 24.04 on arm64 vm pass:
```
Architecture:             aarch64
  CPU op-mode(s):         64-bit
  Byte Order:             Little Endian
```

On benchmark side. Here is a naive try to measure lua vs luajit performance.
First, let's load some large numbers for keys:
```
eval "for i = 1000000, 2000000 do redis.call('SET', i, 1) end" 0
```
Now we would use this script to load the key and check if it's a prime number with not so efficient algorithm:
```
script load "local n = redis.call('randomkey') \n for i = 2, n^(1/2) do if (n % i) == 0 then return false end end return true"
```

Baseline (just randomkey, `valkey-benchmark -n 1000000 randomkey`):
```
  throughput summary: 84875.23 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.318     0.088     0.311     0.439     0.519     1.199
```

Script with Lua:
```
  throughput summary: 53656.71 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.858     0.024     0.823     1.367     1.735     4.767
```

Script with LuaJIT:
```
  throughput summary: 76534.52 requests per second
  latency summary (msec):
          avg       min       p50       p95       p99       max
        0.425     0.040     0.415     0.599     0.735     1.839
```

So, on average case looks like Lua took about `0.858 - 0.318 = 0.54 msec`, and LuaJIT about `0.425 - 0.318 = 0.107 msec`.
In this case we could say that LuaJIT is about 5 times faster than Lua.

**Comment by PingXie** [comment]
I believe we're aligned on the guidelines:

1. Compatibility is essential.
2. Devendoring is preferred.

I have a few questions:

1. What would it take to increase the max string length to 3GB with LuaJIT?
2. Are there any additional known gaps between LuaJIT and the current setup?
3. The patches in [this branch](https://github.com/secwall/valkey/tree/luajit) look non-trivial. I'm curious about the feasibility of upstreaming them. If upstreaming isn't likely, we may face significant work ahead due to vendoring.
4. Are we comfortable with limited support? Specifically, using LuaJIT only for x86/64 and ARM, while other platforms remain on the current setup?

On a related note, we have another discussion around scripting performance improvements via WASM ([issue #387](https://github.com/valkey-io/valkey/issues/387)). It’ll be interesting to compare these approaches: LuaJIT appears more compatible on the surface, but it couldn't achieve full compatibility, while WASM offered a much larger performance boost. The decision may not be straightforward.

@zuiderkwast, is your list below comprehensive? Perhaps we should document it in an RFC with more details. I haven’t come across any standardized benchmarking tests for scripts, so it would be great if we could create one for easier apples-to-apples comparisons in the future, regardless. Thoughts?

> * Small or large values
> * More values, such as multiple keys or multiple fields from hash or set type keys
> * A few lines or many lines of Lua code

**Comment by zuiderkwast** [comment]
> @zuiderkwast, is your list below comprehensive?

@PingXie No, it's arbitrary, but arbitrary benchmarks are better than none. :smile: @secwall's "naive try to measure lua vs luajit performance" and "load the key and check if it's a prime number with not so efficient algorithm" is very good IMO. Now, we have a minimal script using large strings and another script doing some loops and arithmetics, and LuaJIT is at least twice as fast in both. Performance-wise, that's convincing to me.

> Compatibility is essential.

Yeah. But OTOH this has prevented us from updating it for a long time. Now 5.1 is end-of-life since 12 years: *"The last release was [Lua 5.1.5](https://www.lua.org/ftp/lua-5.1.5.tar.gz), released on 17 Feb 2012. There will be no further releases of Lua 5.1."* (https://www.lua.org/versions.html). I'm not sure it's wise to stay on this old version forever... (See also discussion in #1204.)


> 3. The patches in [this branch](https://github.com/secwall/valkey/tree/luajit) look non-trivial. I'm curious about the feasibility of upstreaming them. If upstreaming isn't likely, we may face significant work ahead due to vendoring.
> 4. Are we comfortable with limited support? Specifically, using LuaJIT only for x86/64 and ARM, while other platforms remain on the current setup?

Good questions. I do agree the patches add maintenance burden. About 4, I think it depends if we want to have LuaJIT as an optional alternative, or if we're aiming at replacing the old Lua.

If we have a pluggable scripting engine mechanism, e.g. using the module API, then we could more easily provide alternatives, official and unofficial ones. Running an unpatched LuaJIT can probably be acceptable too for some users. A more modular approach is probably the way forward.

**Comment by secwall** [comment]
> 1. What would it take to increase the max string length to 3GB with LuaJIT?

It would probably take some effort as max buffer in LuaJIT right now is 2GB. Are we sure that 3GB is enough?
Right now in valkey one can do this:
```
127.0.0.1:6379> eval "local f = string.rep(\"x\", 1024 * 1024 * 1024) \n return string.len(f..f..f..f)" 0
(integer) 4294967296
```

I don't think that we need to drop current Lua implementation. A lot of users are migrating from Redis. Breaking compatibility right now would definitely scare them off Valkey.

> 2. Are there any additional known gaps between LuaJIT and the current setup?

Yes, some functions are marked as deprecated in Lua 5.1, but still work. They were removed from LuaJIT. See [here](https://github.com/LuaJIT/LuaJIT/issues/325).

> 3. The patches in [this branch](https://github.com/secwall/valkey/tree/luajit) look non-trivial. I'm curious about the feasibility of upstreaming them. If upstreaming isn't likely, we may face significant work ahead due to vendoring.

We could probably try to make part them not so invasive (don't drop code, just hide it under ifdefs and set something like `-DSANDBOXING` during the build process). But cmsgpack, cjson, cstruct are not part of standard lua library. So it's unlikely that all paches would be accepted.

> 4. Are we comfortable with limited support? Specifically, using LuaJIT only for x86/64 and ARM, while other platforms remain on the current setup?

We could support other archs (`mips`, `mips64`, `arm`, `x86`, and `ppc`). It would just take some time to port readonly tables support for their respective assembly VMs.

**Comment by zuiderkwast** [comment]
> We could probably try to make part them not so invasive (don't drop code, just hide it under ifdefs and set something like -DSANDBOXING during the build process).

Yes, this is needed for upstreaming. Even better if it can be done with in runtime instead of a compile flag, so we can use system installed LuaJIT from distros.

> But cmsgpack, cjson, cstruct are not part of standard lua library. So it's unlikely that all paches would be accepted.

Isn't there some module system, so we can provide them without compiling them into LuaJIT?

> We could support other archs (`mips`, `mips64`, `arm`, `x86`, and `ppc`). It would just take some time to port readonly tables support for their respective assembly VMs.

We should try to do this directly towards the upstream. There is activity on GitHub for LuaJIT development: https://github.com/LuaJIT/LuaJIT/

**Comment by secwall** [comment]
Let's [ask](https://github.com/LuaJIT/LuaJIT/issues/1294) LuaJIT folks about it.

**Comment by zuiderkwast** [comment]
@secwall What so you think of the sandboxin techniques that don't require patching, mentioned in the LuaJIT issue?

**Comment by secwall** [comment]
We should probably get rid of 2 patches out of 3 (load replace to avoid loading bytecode and removing `os.*`, `dofile`, etc). I'll try to do it in a couple of days. And for readonly tables we could probably have separate lua VMs for each user. Would it do the trick?

**Comment by hwware** [comment]
> > Are the benchmarks the same as in #1168 where you work with values of size 100KB?
> 
> Yes, benchmark is the same (large strings, same lua scripts, same way of running `valkey-benchmark`)
> 
> > I believe LuaJIT is faster, but it would be interesting to see some more benchmark comparisons:
> > Small or large values
> > More values, such as multiple keys or multiple fields from hash or set type keys
> > A few lines or many lines of Lua code
> 
> Ok, I'll try to benchmark with these examples. Any other ideas of benchmarks are also welcome.
> 
> > We have some patches to our vendored Lua verions, to make it more sandboxed and safe(?). Do we need to vendor LuaJIT and add similar patches for that? It would be nice if we can avoid patching it and vendoring it.
> 
> I'm not a security expert, but I've tried to sandbox luajit as much as vendored lua is sandboxed. Quick link to patches: https://github.com/secwall/valkey/tree/luajit/deps/luajit/patches. The most interesting from security perspective are 0003, 0005, and 0007. I've mentioned it in first message, but let me clarify it again: only x86_64 is sandboxed for now (because it requires a bit of assembly knowledge to implement readonly tables).

The link https://github.com/secwall/valkey/tree/luajit/deps/luajit/patches shows empty page, Sorry

**Comment by secwall** [comment]
> The link https://github.com/secwall/valkey/tree/luajit/deps/luajit/patches shows empty page, Sorry

You are right, it's broken because Luajit patches are currently moved to another directory in the HEAD of the luajit branch. You might want to take a look [here](https://github.com/secwall/valkey/tree/luajit/deps/luajit-patches), but we decided earlier in this thread to drop all patches for LuaJIT and use the vanilla version. Therefore, it probably doesn't make sense to look there.

Dropping the patches took more time than I expected. And some questions arose:
1) How will the LuaJIT flag work with #1204? Does setting `USE_LUA=no` and `USE_LUAJIT=yes` mean we have neither Lua nor LuaJIT?
2) Creating a Lua VM per user seems secure to me regarding [CVE-2022-24735](https://nvd.nist.gov/vuln/detail/CVE-2022-24735). Am I wrong? Of course, this would use more memory in setups with many users.
3) If we choose to create a VM per user, how should we approach testing? Should we keep tests with the readonly table if LuaJIT isn't used, and skip it if LuaJIT is enabled?

**Comment by zuiderkwast** [comment]
> Dropping the patches took more time than I expected.

So you managed to do all the sandboxing fixes using plain Lua and Lua-C-API calls? This is awesome! Can we do the same with regular Lua and allow building with user-defined Lua, such as 5.4?

> 1. How will the LuaJIT flag work with #1204? Does setting USE_LUA=no and USE_LUAJIT=yes mean we have neither Lua nor LuaJIT?

We will have to make some decisions before we merge any of these features. I hope it will be easy to change this part. Something like `USE_LUA=no | yes | luajit | custom`????

> 2. Creating a Lua VM per user seems secure to me regarding [CVE-2022-24735](https://nvd.nist.gov/vuln/detail/CVE-2022-24735). Am I wrong? Of course, this would use more memory in setups with many users.

Security-wise, I don't know yet. Memory-wise, maybe it can be acceptable. How big is the context? We can create the context lazily the first time a client needs it, so it doesn't consume memory for clients that don't need it.

> 3. If we choose to create a VM per user, how should we approach testing? Should we keep tests with the readonly table if LuaJIT isn't used, and skip it if LuaJIT is enabled?

Yes, something like that can probably be done. For example, we can add some tag to the tests we want to run only with vendored lua, or only with luajit, etc.

**Comment by zuiderkwast** [comment]
We discussed this in an online meeting. We want to be able to extend the scripting languages using the module API.

We want to use the module API to extend the scripting system: #1261

**Comment by secwall** [comment]
Finally I got some time to work on this and made the following: https://github.com/secwall/valkey-luajit
That's a fork of lua module (great thanks to @rjd15372 for working on separation).
It includes FFI interface bridge (disabled by default, thanks to @neomantra for the prototype that I've reused).

Limitations:
1) No read-only tables support - 2 states are used per user for isolation
2) No patches on LuaJIT - so we wrap load to allow text only, and replace some things like os.execute with nil
3) Debug is not working
4) Active defrag gets puzzled on lua memory (because LuaJIT allocator is not compatible with Jemalloc)

~~I had to apply 2 small patches to valkey to make tests working (they are mostly about need debug flag clarification, I could bring them in separate PRs if there is a chance to merge that).~~

UPD: No patches for valkey now.

Overall module seem to work. But I didn't test it throughly so consider it PoC.

**Comment by madolson** [comment]
Decisions from the core meeting. @valkey-io/core-team as an FYI 
1. We will move it in to the Valkey organization.
2. We have a bunch of open questions about how this will eventually get added to the Valkey bundle. So we can pin that for later.

**Comment by dvkashapov** [comment]
Implemented in https://github.com/valkey-io/valkey-luajit
Further development will happen there
